Mathlib/Analysis/Analytic/IsolatedZeros.lean

```diff
@@ -297,5 +307,68 @@ lemma eq_zero_or_eq_zero_of_mul_eq_zero [NoZeroDivisors A]
   eq_zero_or_eq_zero_of_smul_eq_zero hf hg hfg hU
 
 end Mul
-
 end AnalyticOnNhd
+
+/-!
+### Preimages of codiscrete sets
+-/
+
+section PreimgCodiscrete
+
+/-- Preimages of codiscrete sets, local version: if `f` is analytic at `x` and not locally constant,
+then the preimage of any punctured neighbourhood of `f x` is a punctured neighbourhood of `x`. -/
+theorem AnalyticAt.preimg_of_puncNhd {x : 𝕜} {f : 𝕜 → E} {s : Set E} (hfx : AnalyticAt 𝕜 f x)
+    (h₂f : ¬Filter.EventuallyConst f (𝓝 x)) (hs : s ∈ 𝓝[≠] f x) :
+    f ⁻¹' s ∈ 𝓝[≠] x := by
+  have : ∀ᶠ (z : 𝕜) in 𝓝 x, f z ∈ insert (f x) s := by
+    filter_upwards [hfx.continuousAt.preimage_mem_nhds (insert_mem_nhds_iff.2 hs)]
+    tauto
+  by_contra h
+  have : Filter.EventuallyConst f (𝓝 x) := by
+    rw [Filter.eventuallyConst_iff_exists_eventuallyEq]
+    use f x
+    rw [Filter.EventuallyEq, ← hfx.frequently_eq_iff_eventually_eq analyticAt_const]
+    apply ((frequently_imp_distrib_right.2 h).and_eventually
+      (eventually_nhdsWithin_of_eventually_nhds this)).mono
+    intro z ⟨h₁z, h₂z⟩
+    rw [Set.mem_insert_iff] at h₂z
+    tauto
+  tauto
```

Comment:
The `have ... tauto` structure seems a little circuitous. Perhaps `apply h2f.elim`? Also, as Jon already pointed out, the `Filter.xxx` prefixes can be dropped since `Filter` is already opened.

Reply:
Implemented with modifications. I could not get apply `h2f.elim` to work, but I found the tactic `absurd h₂f` which probably does what you suggested, replacing the goal by `EventuallyConst f (𝓝 x)`, which can then be shown without an additional `have … tauto`.

The unnecessary `Filter.…` have been removed.

Thanks for pointing that out.